src/phase0/state_transition/slot_processing.rs

```diff
@@ -67,11 +75,137 @@ pub fn process_slot<
-    state.block_roots[(state.slot % context.slots_per_historical_root as u64) as usize] = previous_block_root;
+    let previous_block_root = state
+        .latest_block_header
+        .hash_tree_root(merkleization_context)?;
+    state.block_roots[(state.slot % context.slots_per_historical_root as u64) as usize] =
+        previous_block_root;
     Ok(())
-}
\ No newline at end of file
+}
+
+pub fn state_transition<
+    const SLOTS_PER_HISTORICAL_ROOT: usize,
+    const HISTORICAL_ROOTS_LIMIT: usize,
+    const ETH1_DATA_VOTES_BOUND: usize,
+    const VALIDATOR_REGISTRY_LIMIT: usize,
+    const EPOCHS_PER_HISTORICAL_VECTOR: usize,
+    const EPOCHS_PER_SLASHINGS_VECTOR: usize,
+    const MAX_VALIDATORS_PER_COMMITTEE: usize,
+    const PENDING_ATTESTATIONS_BOUND: usize,
+    const MAX_PROPOSER_SLASHINGS: usize,
+    const MAX_ATTESTER_SLASHINGS: usize,
+    const MAX_ATTESTATIONS: usize,
+    const MAX_DEPOSITS: usize,
+    const MAX_VOLUNTARY_EXITS: usize,
+>(
+    state: &mut BeaconState<
+        SLOTS_PER_HISTORICAL_ROOT,
+        HISTORICAL_ROOTS_LIMIT,
+        ETH1_DATA_VOTES_BOUND,
+        VALIDATOR_REGISTRY_LIMIT,
+        EPOCHS_PER_HISTORICAL_VECTOR,
+        EPOCHS_PER_SLASHINGS_VECTOR,
+        MAX_VALIDATORS_PER_COMMITTEE,
+        PENDING_ATTESTATIONS_BOUND,
+    >,
+    signed_block: &SignedBeaconBlock<
+        MAX_PROPOSER_SLASHINGS,
+        MAX_VALIDATORS_PER_COMMITTEE,
+        MAX_ATTESTER_SLASHINGS,
+        MAX_ATTESTATIONS,
+        MAX_DEPOSITS,
+        MAX_VOLUNTARY_EXITS,
+    >,
+    validate_result: Option<bool>,
+    context: &Context,
+) -> Result<(), Error> {
+    let merkleization_context = context.for_merkleization();
+    let validate_result = validate_result.unwrap_or(true);
+    let block = &signed_block.message;
+
+    process_slots(state, block.slot, context)?;
+    if validate_result {
+        verify_block_signature(state, signed_block, context)?;
+    }
+    process_block(state, block, context)?;
+    if validate_result && block.state_root != state.hash_tree_root(merkleization_context)? {
+        return Err(InvalidStateRoot);
+    }
+
+    Ok(())
+}
+
+pub fn get_matching_source_attestations<
+    const SLOTS_PER_HISTORICAL_ROOT: usize,
+    const HISTORICAL_ROOTS_LIMIT: usize,
+    const ETH1_DATA_VOTES_BOUND: usize,
+    const VALIDATOR_REGISTRY_LIMIT: usize,
+    const EPOCHS_PER_HISTORICAL_VECTOR: usize,
+    const EPOCHS_PER_SLASHINGS_VECTOR: usize,
+    const MAX_VALIDATORS_PER_COMMITTEE: usize,
+    const PENDING_ATTESTATIONS_BOUND: usize,
+>(
+    state: &BeaconState<
+        SLOTS_PER_HISTORICAL_ROOT,
+        HISTORICAL_ROOTS_LIMIT,
+        ETH1_DATA_VOTES_BOUND,
+        VALIDATOR_REGISTRY_LIMIT,
+        EPOCHS_PER_HISTORICAL_VECTOR,
+        EPOCHS_PER_SLASHINGS_VECTOR,
+        MAX_VALIDATORS_PER_COMMITTEE,
+        PENDING_ATTESTATIONS_BOUND,
+    >,
+    epoch: Epoch,
+    context: &Context,
+) -> Result<Vec<PendingAttestation<MAX_VALIDATORS_PER_COMMITTEE>>, Error> {
```

Comment:
something to consider if we start racking up very nested types is to define a type alias (next to `state_transition::Error`), something like:

```rust
type StateTransitionResult<T> = Result<T, Error>;
```

Reply:
then we can just write here:

```rust
StateTransitionResult<Vec<PendingAttestation<MAX_VALIDATORS_PER_COMMITTEE>>>
```